modules/library/render/src/main/java/org/geotools/renderer/lite/gridcoverage2d/GridCoverageRenderer.java

```diff
@@ -970,6 +976,38 @@ private RenderedImage renderImage(final GridCoverage2DReader reader,
             }
         }
         
+        // Check if reader supports band selection, and rearrange raster channels order in
+        // symbolizer. Reader should have taken care o proper channel order, based on initial
+        // symbolizer channel definition
+        if (reader.getFormat().getReadParameters().getDescriptor().descriptors().
+        		contains(AbstractGridFormat.BANDS) && readParams!=null){
+	        // if bands are selected, alter the symbolizer to use bands in order 0,1,2,... since the
+	        // channel order defined by it previously is taken care of the reader
+            int[] bandIndices = null;
+        	for (int i = 0; i < readParams.length; i++) {
+                final ParameterValue param = (ParameterValue) readParams[i];
+                if (param.getDescriptor().equals(AbstractGridFormat.BANDS)) {
+                    bandIndices = (int[]) param.getValue();
+	                break;
+                }
+            }
+	        if (bandIndices!=null){
+	            ChannelSelection selection = symbolizer.getChannelSelection();
+	            final SelectedChannelType[] channels = selection.getSelectedChannels();
+	            if (channels!=null){
+		            int i=0;
+		            for (SelectedChannelType channel:channels){
+		                //Remember, channel indices start from 1
+		                channel.setChannelName(Integer.toString(i+1));
+		                i++;
+		            }
+
+		            ChannelSelectionUpdateStyleVisitor channelsUpdateVisitor = new ChannelSelectionUpdateStyleVisitor(channels);
+	                symbolizer.accept(channelsUpdateVisitor);
```

Comment:
By itself this does nothing. The proper way to use a duplicating visitor is to make it duplicate the entire symbolizer, and replace the old symbolizer by the one returned by the visit. The above code modifying in place the old symbolizer can cause trouble as styles can be cached by the streaming renderer caller.